The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11166569131/job/31040861719) [(plain)](https://github.com/rust-lang/rust/commit/34210ca25b1235049e4b6e740882942b84cdadfc/checks/31040861719/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:20950f2ccee3dff53a038adf5c1cf05231c0b30772617126a5f6478a66316a29cba9aead69f7bb0004886d32c1f8e6287542cf6d25130711c82d16a66201d4fe:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
failures:

---- [codegen] tests/codegen/i128-x86-align.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-18/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/i128-x86-align/i128-x86-align.ll" "/checkout/tests/codegen/i128-x86-align.rs" "--check-prefix=CHECK" "--check-prefix" "NONMSVC" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
/checkout/tests/codegen/i128-x86-align.rs:26:17: error: CHECK-NEXT: expected string not found in input
/checkout/tests/codegen/i128-x86-align.rs:26:17: error: CHECK-NEXT: expected string not found in input
 // CHECK-NEXT: [[IV1:%.*]] = insertvalue { i32, i128 } poison, i32 [[A]], 0
                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/i128-x86-align/i128-x86-align.ll:13:34: note: scanning from here
 %2 = load i128, ptr %1, align 16, !noundef !3
                                 ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/i128-x86-align/i128-x86-align.ll:13:34: note: with "A" equal to "%0"
 %2 = load i128, ptr %1, align 16, !noundef !3
                                 ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/i128-x86-align/i128-x86-align.ll:15:13: note: possible intended match here
 %3 = getelementptr inbounds i8, ptr %_0, i64 16
/checkout/tests/codegen/i128-x86-align.rs:68:18: error: CHECK-LABEL: expected string not found in input
/checkout/tests/codegen/i128-x86-align.rs:68:18: error: CHECK-LABEL: expected string not found in input
 // CHECK-LABEL: define { i128, i128 } @transmute(i32 noundef %x.0, i128 noundef %x.1)
                 ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/i128-x86-align/i128-x86-align.ll:43:28: note: scanning from here
define void @load_volatile(ptr dead_on_unwind noalias nocapture noundef writable sret([32 x i8]) align 16 dereferenceable(32) %_0, ptr noalias noundef readonly align 16 dereferenceable(32) %x) unnamed_addr #0 {
                           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/i128-x86-align/i128-x86-align.ll:51:95: note: possible intended match here
define void @transmute(ptr dead_on_unwind noalias nocapture noundef writable sret([32 x i8]) align 16 dereferenceable(32) %_0, i32 noundef %x.0, i128 noundef %x.1) unnamed_addr #0 {

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/i128-x86-align/i128-x86-align.ll
Check file: /checkout/tests/codegen/i128-x86-align.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'i128_x86_align.c99a19dc19580418-cgu.0' 
            2: source_filename = "i128_x86_align.c99a19dc19580418-cgu.0" 
            3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
            4: target triple = "x86_64-unknown-linux-gnu" 
            5:  
            6: %ScalarPair = type { i32, [3 x i32], i128 } 
            7:  
            8: ; Function Attrs: nonlazybind uwtable 
            9: define void @load(ptr dead_on_unwind noalias nocapture noundef writable sret([32 x i8]) align 16 dereferenceable(32) %_0, ptr noalias noundef readonly align 16 dereferenceable(32) %x) unnamed_addr #0 { 
           10: start: 
           11:  %0 = load i32, ptr %x, align 16, !noundef !3 
           12:  %1 = getelementptr inbounds i8, ptr %x, i64 16 
           13:  %2 = load i128, ptr %1, align 16, !noundef !3 
next:26'0                                       X~~~~~~~~~~~~~ error: no match found
next:26'1                                                      with "A" equal to "%0"
           14:  store i32 %0, ptr %_0, align 16 
next:26'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           15:  %3 = getelementptr inbounds i8, ptr %_0, i64 16 
next:26'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
next:26'2                  ?                                     possible intended match
           16:  store i128 %2, ptr %3, align 16 
next:26'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           17:  ret void 
next:26'0      ~~~~~~~~~~
           18: } 
next:26'0      ~~
next:26'0      ~
next:26'0      ~
           20: ; Function Attrs: nonlazybind uwtable 
next:26'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           21: define void @store(ptr noalias noundef align 16 dereferenceable(32) %x) unnamed_addr #0 { 
next:26'0      ~~~~~~~~~~~~~~~~~~~
           22: start: 
           23:  store i32 1, ptr %x, align 16 
           24:  %0 = getelementptr inbounds i8, ptr %x, i64 16 
           25:  store i128 2, ptr %0, align 16 
           26:  ret void 
           27: } 
           28:  
           29: ; Function Attrs: nonlazybind uwtable 
           30: define void @alloca() unnamed_addr #0 { 
           31: start: 
           32:  %x = alloca [32 x i8], align 16 
           33:  call void @llvm.lifetime.start.p0(i64 32, ptr %x) 
           34:  store i32 1, ptr %x, align 16 
           35:  %0 = getelementptr inbounds i8, ptr %x, i64 16 
           36:  store i128 2, ptr %0, align 16 
           37:  call void @store(ptr noalias noundef align 16 dereferenceable(32) %x) 
           38:  call void @llvm.lifetime.end.p0(i64 32, ptr %x) 
           39:  ret void 
           40: } 
           41:  
           42: ; Function Attrs: nonlazybind uwtable 
           43: define void @load_volatile(ptr dead_on_unwind noalias nocapture noundef writable sret([32 x i8]) align 16 dereferenceable(32) %_0, ptr noalias noundef readonly align 16 dereferenceable(32) %x) unnamed_addr #0 { 
label:68'0                                X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
label:68'0     ~~~~~~~
label:68'0     ~~~~~~~
           45:  %0 = load volatile %ScalarPair, ptr %x, align 16 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           46:  store %ScalarPair %0, ptr %_0, align 16 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           47:  ret void 
label:68'0     ~~~~~~~~~~
           48: } 
label:68'0     ~~
label:68'0     ~
label:68'0     ~
           50: ; Function Attrs: nonlazybind uwtable 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           51: define void @transmute(ptr dead_on_unwind noalias nocapture noundef writable sret([32 x i8]) align 16 dereferenceable(32) %_0, i32 noundef %x.0, i128 noundef %x.1) unnamed_addr #0 { 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
label:68'1                                                                                                   ?                                                                                        possible intended match
label:68'0     ~~~~~~~
label:68'0     ~~~~~~~
           53:  store i32 %x.0, ptr %_0, align 16 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           54:  %0 = getelementptr inbounds i8, ptr %_0, i64 16 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           55:  store i128 %x.1, ptr %0, align 16 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           56:  ret void 
label:68'0     ~~~~~~~~~~
           57: } 
label:68'0     ~~
label:68'0     ~
label:68'0     ~
           59: ; Function Attrs: nonlazybind uwtable 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           60: define void @store_struct(ptr noalias noundef align 16 dereferenceable(32) %x) unnamed_addr #0 { 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
label:68'0     ~~~~~~~
           62:  %_2 = alloca [32 x i8], align 16 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           63:  call void @llvm.lifetime.start.p0(i64 32, ptr %_2) 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           64:  store i32 1, ptr %_2, align 16 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           65:  %0 = getelementptr inbounds i8, ptr %_2, i64 4 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           66:  store i32 2, ptr %0, align 4 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           67:  %1 = getelementptr inbounds i8, ptr %_2, i64 16 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           68:  store i128 3, ptr %1, align 16 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           69:  call void @llvm.memcpy.p0.p0.i64(ptr align 16 %x, ptr align 16 %_2, i64 32, i1 false) 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           70:  call void @llvm.lifetime.end.p0(i64 32, ptr %_2) 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           71:  ret void 
label:68'0     ~~~~~~~~~~
           72: } 
label:68'0     ~~
label:68'0     ~
label:68'0     ~
           74: ; Function Attrs: nocallback nofree nounwind willreturn memory(argmem: readwrite) 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           75: declare void @llvm.memcpy.p0.p0.i64(ptr noalias nocapture writeonly, ptr noalias nocapture readonly, i64, i1 immarg) #1 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
label:68'0     ~
label:68'0     ~
           77: ; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           78: declare void @llvm.lifetime.start.p0(i64 immarg, ptr nocapture) #2 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
label:68'0     ~
label:68'0     ~
           80: ; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           81: declare void @llvm.lifetime.end.p0(i64 immarg, ptr nocapture) #2 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
label:68'0     ~
label:68'0     ~
           83: attributes #0 = { nonlazybind uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" } 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           84: attributes #1 = { nocallback nofree nounwind willreturn memory(argmem: readwrite) } 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           85: attributes #2 = { nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) } 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
label:68'0     ~
label:68'0     ~
           87: !llvm.module.flags = !{!0, !1} 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           88: !llvm.ident = !{!2} 
label:68'0     ~~~~~~~~~~~~~~~~~~~~
label:68'0     ~
label:68'0     ~
           90: !0 = !{i32 8, !"PIC Level", i32 2} 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           91: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           92: !2 = !{!"rustc version 1.83.0-nightly (4c6421a22 2024-10-03)"} 
label:68'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           93: !3 = !{} 
label:68'0     ~~~~~~~~~
------------------------------------------



```

</details>

--- a/compiler/rustc_ty_utils/src/abi.rs
+++ b/compiler/rustc_ty_utils/src/abi.rs
@@ -726,6 +726,49 @@ fn fn_abi_adjust_for_abi<'tcx>(
                 };
             }
 
+            if arg_idx.is_none() && arg.layout.size > Pointer(AddressSpace::DATA).size(cx) * 2 {
+                // Return values larger than 2 registers using a return area
+                // pointer. LLVM and Cranelift disagree about how to return
+                // values that don't fit in the registers designated for return
+                // values. LLVM will force the entire return value to be passed
+                // by return area pointer, while Cranelift will look at each IR level
+                // return value independently and decide to pass it in a
+                // register or not, which would result in the return value
+                // being passed partially in registers and partially through a
+                // return area pointer.
+                //
+                // While Cranelift may need to be fixed as the LLVM behavior is
+                // generally more correct with respect to the surface language,
+                // forcing this behavior in rustc itself makes it easier for
+                // other backends to conform to the Rust ABI and for the C ABI
+                // rustc already handles this behavior anyway.
+                //
+                // In addition LLVM's decision to pass the return value in
+                // registers or using a return area pointer depends on how
+                // exactly the return type is lowered to an LLVM IR type. For
+                // example `Option<u128>` can be lowered as `{ i128, i128 }`
+                // in which case the x86_64 backend would use a return area
+                // pointer, or it could be passed as `{ i32, i128 }` in which
+                // case the x86_64 backend would pass it in registers by taking
+                // advantage of an LLVM ABI extension that allows using 3
+                // registers for the x86_64 sysv call conv rather than the
+                // officially specified 2 registers.
+                //
+                // FIXME: Technically we should look at the amount of available
+                // return registers rather than guessing that there are 2
+                // registers for return values. In practice only a couple of
+                // architectures have less than 2 return registers. None of
+                // which supported by Cranelift.
+                //
+                // NOTE: This adjustment is only necessary for the Rust ABI as
+                // for other ABI's the calling convention implementations in
+                // rustc_target already ensure any return value which doesn't
+                // fit in the available amount of return registers is passed in
+                // the right way for the current target.
+                arg.make_indirect();
+                return;
+            }
+
             match arg.layout.abi {
                 Abi::Aggregate { .. } => {}
 

--- a/tests/codegen/i128-x86-align.rs
+++ b/tests/codegen/i128-x86-align.rs
@@ -19,13 +19,15 @@ pub struct ScalarPair {
 #[no_mangle]
 pub fn load(x: &ScalarPair) -> ScalarPair {
     // CHECK-LABEL: @load(
+    // CHECK-SAME: sret([32 x i8]) align 16 dereferenceable(32) %_0,
     // CHECK-SAME: align 16 dereferenceable(32) %x
     // CHECK:      [[A:%.*]] = load i32, ptr %x, align 16
     // CHECK-NEXT: [[GEP:%.*]] = getelementptr inbounds i8, ptr %x, i64 16
     // CHECK-NEXT: [[B:%.*]] = load i128, ptr [[GEP]], align 16
-    // CHECK-NEXT: [[IV1:%.*]] = insertvalue { i32, i128 } poison, i32 [[A]], 0
-    // CHECK-NEXT: [[IV2:%.*]] = insertvalue { i32, i128 } [[IV1]], i128 [[B]], 1
-    // CHECK-NEXT: ret { i32, i128 } [[IV2]]
+    // CHECK-NEXT: store i32 [[A]], ptr %_0, align 16
+    // CHECK-NEXT: [[GEP:%.*]] = getelementptr inbounds i8, ptr %_0, i64 16
+    // CHECK-NEXT: store i128 [[B]], ptr [[GEP]], align 16
+    // CHECK-NEXT: ret void
     *x
 }
 
@@ -53,29 +55,23 @@ pub fn alloca() {
 #[no_mangle]
 pub fn load_volatile(x: &ScalarPair) -> ScalarPair {
     // CHECK-LABEL: @load_volatile(
+    // CHECK-SAME: sret([32 x i8]) align 16 dereferenceable(32) %_0,
     // CHECK-SAME: align 16 dereferenceable(32) %x
-    // CHECK:      [[TMP:%.*]] = alloca [32 x i8], align 16
     // CHECK:      [[LOAD:%.*]] = load volatile %ScalarPair, ptr %x, align 16
-    // CHECK-NEXT: store %ScalarPair [[LOAD]], ptr [[TMP]], align 16
-    // CHECK-NEXT: [[A:%.*]] = load i32, ptr [[TMP]], align 16
-    // CHECK-NEXT: [[GEP:%.*]] = getelementptr inbounds i8, ptr [[TMP]], i64 16
-    // CHECK-NEXT: [[B:%.*]] = load i128, ptr [[GEP]], align 16
+    // CHECK-NEXT: store %ScalarPair [[LOAD]], ptr %_0, align 16
+    // CHECK-NEXT: ret void
     unsafe { std::intrinsics::volatile_load(x) }
 }
 
 #[no_mangle]
 pub fn transmute(x: ScalarPair) -> (std::mem::MaybeUninit<i128>, i128) {
-    // CHECK-LABEL: define { i128, i128 } @transmute(i32 noundef %x.0, i128 noundef %x.1)
-    // CHECK:       [[TMP:%.*]] = alloca [32 x i8], align 16
-    // CHECK-NEXT:  store i32 %x.0, ptr [[TMP]], align 16
-    // CHECK-NEXT:  [[GEP:%.*]] = getelementptr inbounds i8, ptr [[TMP]], i64 16
+    // CHECK-LABEL: @transmute(
+    // CHECK-SAME:  sret([32 x i8]) align 16 dereferenceable(32) %_0,
+    // CHECK-SAME:  i32 noundef %x.0, i128 noundef %x.1
+    // CHECK:       store i32 %x.0, ptr %_0, align 16
+    // CHECK-NEXT:  [[GEP:%.*]] = getelementptr inbounds i8, ptr %_0, i64 16
     // CHECK-NEXT:  store i128 %x.1, ptr [[GEP]], align 16
-    // CHECK-NEXT:  [[LOAD1:%.*]] = load i128, ptr %_0, align 16
-    // CHECK-NEXT:  [[GEP2:%.*]] = getelementptr inbounds i8, ptr [[TMP]], i64 16
-    // CHECK-NEXT:  [[LOAD2:%.*]] = load i128, ptr [[GEP2]], align 16
-    // CHECK-NEXT:  [[IV1:%.*]] = insertvalue { i128, i128 } poison, i128 [[LOAD1]], 0
-    // CHECK-NEXT:  [[IV2:%.*]] = insertvalue { i128, i128 } [[IV1]], i128 [[LOAD2]], 1
-    // CHECK-NEXT:  ret { i128, i128 } [[IV2]]
+    // CHECK-NEXT:  ret void
     unsafe { std::mem::transmute(x) }
 }
 

--- a/tests/codegen/tuple-layout-opt.rs
+++ b/tests/codegen/tuple-layout-opt.rs
@@ -19,28 +19,28 @@ pub fn test_ScalarZstFirst(_: ScalarZstFirst) -> ScalarZstFirst {
 }
 
 type ScalarPairZstLast = (u8, u128, ());
-// CHECK: define {{(dso_local )?}}{ i128, i8 } @test_ScalarPairZstLast(i128 %_1.0, i8 %_1.1)
+// CHECK: define {{(dso_local )?}}void @test_ScalarPairZstLast(ptr sret([32 x i8]) align 16 %_0, i128 %_1.0, i8 %_1.1)
 #[no_mangle]
 pub fn test_ScalarPairZstLast(_: ScalarPairZstLast) -> ScalarPairZstLast {
     loop {}
 }
 
 type ScalarPairZstFirst = ((), u8, u128);
-// CHECK: define {{(dso_local )?}}{ i8, i128 } @test_ScalarPairZstFirst(i8 %_1.0, i128 %_1.1)
+// CHECK: define {{(dso_local )?}}void @test_ScalarPairZstFirst(ptr sret([32 x i8]) align 16 %_0, i8 %_1.0, i128 %_1.1)
 #[no_mangle]
 pub fn test_ScalarPairZstFirst(_: ScalarPairZstFirst) -> ScalarPairZstFirst {
     loop {}
 }
 
 type ScalarPairLotsOfZsts = ((), u8, (), u128, ());
-// CHECK: define {{(dso_local )?}}{ i128, i8 } @test_ScalarPairLotsOfZsts(i128 %_1.0, i8 %_1.1)
+// CHECK: define {{(dso_local )?}}void @test_ScalarPairLotsOfZsts(ptr sret([32 x i8]) align 16 %_0, i128 %_1.0, i8 %_1.1)
 #[no_mangle]
 pub fn test_ScalarPairLotsOfZsts(_: ScalarPairLotsOfZsts) -> ScalarPairLotsOfZsts {
     loop {}
 }
 
 type ScalarPairLottaNesting = (((), ((), u8, (), u128, ())), ());
-// CHECK: define {{(dso_local )?}}{ i128, i8 } @test_ScalarPairLottaNesting(i128 %_1.0, i8 %_1.1)
+// CHECK: define {{(dso_local )?}}void @test_ScalarPairLottaNesting(ptr sret([32 x i8]) align 16 %_0, i128 %_1.0, i8 %_1.1)
 #[no_mangle]
 pub fn test_ScalarPairLottaNesting(_: ScalarPairLottaNesting) -> ScalarPairLottaNesting {
     loop {}

